allennlp/training/trainer.py

```diff
@@ -298,13 +306,21 @@ def _metrics_to_console(self,  # pylint: disable=no-self-use
         if val_metrics:
             message_template = "Training %s : %3f    Validation %s : %3f "
         else:
-            message_template = "Training %s : %3f "
-
-        for name, value in train_metrics.items():
-            if val_metrics:
-                logger.info(message_template, name, value, name, val_metrics[name])
-            else:
-                logger.info(message_template, name, value)
+            message_template = "%s %s : %3f "
+        
+        metric_names = set(train_metrics.keys())
+        if val_metrics:
+            metric_names.update(val_metrics.keys())
+
+        for name in metric_names:
+            train_metric = train_metrics.get(name, None)
+            val_metric = val_metrics.get(name, None)
+            if val_metric is not None and train_metric is not None:
+                logger.info(message_template, name, train_metric, name, val_metric)
+            elif val_metric is not None:
+                logger.info(message_template, "Validation", name, val_metric)
+            elif train_metric is not None:
+                logger.info(message_template, "Training", name, train_metric)
```

Comment:
I'm not convinced this does the right thing

the `message_template` is chosen based on whether there are any validation metrics or not, but aren't we trying to address the case where there is a validation metric that isn't a train metric? in that case we'd have the first message template, and then this logger statement isn't right?